fix(ast/estree): fix `StringLiteral`s containing lone surrogates

overlookmotel opened the pull request:
Improve output for `StringLiteral`s containing lone surrogates.

Most of the remaining failing tests are just due to difference in case e.g.:

```diff
- "value": "\ud83d",
+ "value": "\uD83D",
```

But there are still a few genuine test fails. We don't have it exactly right yet. But this is a step in the right direction.

hi-ogawa commented on crates/oxc_ast/src/serialize.rs:
```diff
@@ -225,6 +227,71 @@ impl ESTree for NullLiteralRaw<'_> {
+            // because they're already escaped.
+            // We could make this more performant, but it should be vanishingly rare to hit this path anyway.
+            let mut encoded = Vec::with_capacity(raw_unquoted.len());
+            let mut has_slash = false;
+            for &b in raw_unquoted.as_bytes() {
+                if b == b'\\' {
+                    has_slash = !has_slash;
+                } else if b == b'"' && !has_slash {
+                    encoded.push(b'\\');
+                }
+                encoded.push(b);
+            }
```
Is `has_slash` correct? It looks like it's counting all previous occurrence instead of preceding sequence. 
For example, `'\uD83D"'` is edge case and it doesn't escape to `"\uD83D\""`?

(I know this is extremely rare edge cases, but I wonder if we can add extra manual test case since test262 didn't have this one).

overlookmotel replied:
You're right! I neglected to reset `has_slash = false` after a non-slash character. Fixed now.

Anyway, this is superceded by #10041.

As far as tests, it would be a pain to add an ESTree test, because it'd complicate the test runner. But we can add parser tests in #10041. There are some other new tests we need for that PR anyway.

overlookmotel committed:
fix(ast/estree): fix `StringLiteral`s containing lone surrogates (#10036)

Improve output for `StringLiteral`s containing lone surrogates.

Most of the remaining failing tests are just due to difference in case e.g.:

```diff
- "value": "\ud83d",
+ "value": "\uD83D",
```

But there are still a few genuine test fails. We don't have it exactly right yet. But this is a step in the right direction.
